### PR TITLE
Custom Commands For Creating Dummy Entries

### PR DESCRIPTION
**Pull request by saadali5997**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -8,6 +8,7 @@ for line in sys.stdin:
 		print("%-20s %s" % (target, help))
 endef
 export PRINT_HELP_PYSCRIPT
+total = 10
```
Comment (Ayub-Khan) [inline Makefile]: 'CONSTANT'

File: client/management/commands/create_dummy_clients.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+from django.core.management.base import BaseCommand, CommandError
+
+from client.models import Client
+from tms.utils import RandomGenerator
+
+
+class Command(BaseCommand, RandomGenerator):
+    """Create dummy users class."""
+
+    help = "Create fifteen dummy users for testing purposes."
```
Comment (Ayub-Khan) [inline client/management/commands/create_dummy_clients.py]: fifteen?

File: client/tests/test_client.py [hunk 3]
```diff
@@ -85,3 +88,11 @@ def test_delete_clients(self):
         self.assertEqual(response.status_code, 302)
         self.assertEqual(Client.objects.count(), 0)
         self.api_client.logout()
+
+    def test_create_dummy_clients(self):
+        """Test created dummy clients command."""
+        out = StringIO()
+        # args = []
+        # opts = {'total': 15}
+        call_command('create_dummy_clients', 10, stdout=out)
```
Comment (saadali5997) [inline client/tests/test_client.py]: @Ayub-Khan can you help me out a little here?
The command `python manage.py create_dummy_clients 10` works fine on terminal.
But when I call from here like in the line or like this
```
call_command('create_dummy_clients', total=10, stdout=out)
```
Or even like this
```
call_command('create_dummy_clients', '--total=10', stdout=out)
```
it fails. Do you have any sample project where you have called the command in a test case?

File: client/management/commands/create_dummy_clients.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        """Command operations."""
+        no_of_clients = kwargs.get('total', 15)
+
+        name_set = list(self.get_n_unique_names(no_of_clients))
+        email_set = list(self.get_n_unique_emails(no_of_clients))
+        phone_set = list(self.get_n_unique_numbers(no_of_clients))
+        client_list = []
+        i = 0
+
+        while i < no_of_clients:
```
Comment (Ayub-Khan) [inline client/management/commands/create_dummy_clients.py]: why while ?

File: Makefile [hunk 5]
```diff
@@ -49,4 +50,17 @@ requirements: ## installs all dependencies
...
\ No newline at end of file
+	python manage.py runserver
+
+create-dummy-data: ## create dummy entries in models.
+ifdef $(TOTAL)
+	python manage.py create_dummy_clients --total $(TOTAL)
+	python manage.py create_dummy_employees --total $(TOTAL)
+	python manage.py create_dummy_orders --total $(TOTAL)
+	python manage.py create_dummy_tasks --total $(TOTAL)
+else
```
Comment (Ayub-Khan) [inline Makefile:61]: No need for if else i think. but if you wanna keep it up to you.

File: client/management/commands/create_dummy_clients.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+class Command(BaseCommand, RandomGenerator):
+    """Create dummy clients command class."""
+
+    help = "Create dummy clients for testing purposes."
+
+    def add_arguments(self, parser):
+        """Argument specifying how many dummy clients to create."""
+        parser.add_argument(
+            '--total', type=int, help='Indicates the number of clients to be created.', default=5,
```
Comment (Ayub-Khan) [inline client/management/commands/create_dummy_clients.py]: 5 too short. 10-15.

File: client/management/commands/create_dummy_clients.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            self.stdout.write('Can not create -ive or zero number of clients.')
+
+        else:
+            name_set = list(self.get_n_unique_names(no_of_clients))
+            email_set = list(self.get_n_unique_emails(no_of_clients))
+            phone_set = list(self.get_n_unique_numbers(no_of_clients))
+
+            client_list = []
+
+            for i in range(0, no_of_clients):
```
Comment (Ayub-Khan) [inline client/management/commands/create_dummy_clients.py]: no need to say 0.

File: client/tests/test_client.py [hunk 8]
```diff
@@ -85,3 +89,15 @@ def test_delete_clients(self):
         self.assertEqual(response.status_code, 302)
         self.assertEqual(Client.objects.count(), 0)
         self.api_client.logout()
+
+    @ddt.data(
+        (0, 'Can not create -ive or zero number of clients.\n'),
+        (-1, 'Can not create -ive or zero number of clients.\n'),
+        (100, 'Dummy clients created successfully.\n')
+    )
+    @ddt.unpack
```
Comment (Ayub-Khan) [inline client/tests/test_client.py]: Those are fine here but its a good practice to create commands tests underst management tests module.

File: client/management/commands/create_dummy_clients.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
+"""Create dummy clients for testing purpose."""
```
Comment (Ayub-Khan) [inline client/management/commands/create_dummy_clients.py:1]: We need a command to delete dummy data too.

File: client/tests/test_client.py [hunk 10]
```diff
@@ -85,3 +89,15 @@ def test_delete_clients(self):
         self.assertEqual(response.status_code, 302)
         self.assertEqual(Client.objects.count(), 0)
         self.api_client.logout()
+
+    @ddt.data(
+        (0, 'Can not create -ive or zero number of clients.\n'),
```
Comment (Ayub-Khan) [inline client/tests/test_client.py]: no need for \n, you can use assert in.

File: tms/utils.py [hunk 11]
```diff
@@ -1,21 +1,77 @@
...
 from product.models import Product, ProductImages
 
 
-class TestDbSetUp:
+def get_future_date(days_to_add=5):
+    """Return date in future."""
+    return datetime.today() + timedelta(days=days_to_add)
+
+
+class RandomGenerator:
```
Comment (Ayub-Khan) [inline tms/utils.py]: RandomDataGenerator

Comment (Ayub-Khan) [reply tms/utils.py]: random generator usually means random numbers.

File: tms/utils.py [hunk 12]
```diff
@@ -1,21 +1,77 @@
...
+    """Functions for creating random strings."""
+
+    def get_random_alpha_string(self, length=5):
+        """Generate a random alpha string."""
+        random_string = ''.join(random.choices(string.ascii_letters, k=length))
```
Comment (Ayub-Khan) [inline tms/utils.py:34]: you might wanna add spaces?

File: tms/utils.py [hunk 13]
```diff
@@ -1,21 +1,77 @@
...
+                name_set.add(name)
+                i += 1
+        return name_set
+
+    def get_n_unique_emails(self, n):
+        """List of emails."""
+        i = 0
+        email_set = set()
+        while True and i < n:
+            email = self.get_random_alpha_string() + '@test.com'
```
Comment (Ayub-Khan) [inline tms/utils.py]: create a DUMMY_DATA_MARKER constant in management command and that should send the value '@dummytestvalues.com'

File: client/management/commands/create_dummy_clients.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+    def handle(self, *args, **kwargs):
+        """Create 'total' dummy clients in database."""
+        no_of_clients = kwargs['total']
+
+        if no_of_clients < 1:
+            self.stdout.write('Can not create -ive or zero number of clients.')
+
+        else:
+            name_set = list(self.get_n_unique_names(no_of_clients))
```
Comment (Ayub-Khan) [inline client/management/commands/create_dummy_clients.py]: no need to typecast here you can typecast in data generator.

File: tms/utils.py [hunk 15]
```diff
@@ -1,21 +1,77 @@
...
+    def get_n_unique_names(self, n):
+        """List of space separated title case strings."""
+        i = 0
+        name_set = set()
+        while True and i < n:
+            name = self.get_random_alpha_string().title() + ' ' + self.get_random_alpha_string().title()
+            if name not in name_set:
+                name_set.add(name)
+                i += 1
+        return name_set
```
Comment (Ayub-Khan) [inline tms/utils.py:51]: you can type cast to list here.

**Review by Ayub-Khan — CHANGES_REQUESTED** [review]

File: employee/management/commands/create_dummy_employees.py [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+        """Create 'total' dummy employees in database."""
+        no_of_employees = kwargs['total']
+
+        if no_of_employees < 1:
+            self.stdout.write("Can not create -ive or zero number of employees.")
+
+        else:
+            name_set = self.get_n_unique_names(no_of_employees)
+
+            existing_phones_dict = Employee.objects.all().values('phone_number')
```
Comment (Ayub-Khan) [inline employee/management/commands/create_dummy_employees.py]: https://docs.djangoproject.com/en/dev/ref/models/querysets/#values-list

File: employee/management/commands/create_dummy_employees.py [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+
+            for i in range(no_of_employees):
+                employee = Employee(**{
+                    'name': name_set[i],
+                    'gender': 'M',
+                    'phone_number': phone_set[i],
+                    'address': DUMMY_ADDRESS_MARKER
+                })
+                employee_list.append(employee)
+                i += 1
```
Comment (Ayub-Khan) [inline employee/management/commands/create_dummy_employees.py]: no need for i+=1

File: employee/management/tests/test_create_dummy_employees.py [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+from tms import constants
+from tms.utils import TestDbSetUp
+
+
+@ddt.ddt
+class CreateDummyClientsTest(TestCase, TestDbSetUp):
+    """Dummy employees creation test class."""
+
+    def setUp(self):
+        """Create a dummy user for the purpose of testing."""
```
Comment (Ayub-Khan) [inline employee/management/tests/test_create_dummy_employees.py]: This is not used.

File: client/management/tests/test_create_dummy_clients.py [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+from tms import constants
+from tms.utils import TestDbSetUp
+
+
+@ddt.ddt
+class CreateDummyClientsTest(TestCase, TestDbSetUp):
+    """Dummy clients creation test class."""
+
+    def setUp(self):
+        """Create a dummy user for the purpose of testing."""
```
Comment (Ayub-Khan) [inline client/management/tests/test_create_dummy_clients.py]: No need of setup.

File: employee/management/tests/test_delete_dummy_employees.py [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+from tms.constants import DUMMY_EMAIL_MARKER
+from tms.utils import TestDbSetUp
+
+
+class DeleteDummyEmployeesTest(TestCase, TestDbSetUp):
+    """Dummy employees deletion test class."""
+
+    def setUp(self):
+        """Setup employee for deletion."""
+        self.employee = self.create_employee()
```
Comment (Ayub-Khan) [inline employee/management/tests/test_delete_dummy_employees.py:18]: why not sending marker here ?

File: employee/tests.py [hunk 21]
```diff
@@ -54,3 +58,15 @@ def test_delete_employee(self):
         response = self.api_client.post(path, data)
         self.assertEqual(response.status_code, 302)
         self.assertFalse(Employee.objects.filter(id=self.employee.id).exists())
+
+    @ddt.data(
+        (0, 'Can not create -ive or zero number of employees.\n'),
```
Comment (Ayub-Khan) [inline employee/tests.py]: code duplication need to be removed as we have added seperate tests file.

File: order/management/tests/test_create_dummy_tasks.py [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+@ddt.ddt
+class CreateDummyTasksTest(TestCase, TestDbSetUp):
+    """Dummy tasks creation test class."""
+
+    def setUp(self):
+        """Create a dummy user for the purpose of testing."""
+        self.user, self.username, self.password = self.create_user()
+        self.client = self.create_client(email_post_fix=DUMMY_EMAIL_MARKER)
+        self.order = self.create_order(self.client)
+        _ = self.create_order(self.client)
```
Comment (Ayub-Khan) [inline order/management/tests/test_create_dummy_tasks.py]: no need for _ =

File: order/management/tests/test_create_dummy_tasks.py [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+
+    @ddt.data(
+        (0, 'Can not create -ive or zero number of tasks against 0 or no orders.'),
+        (-1, 'Can not create -ive or zero number of tasks against 0 or no orders.'),
+        (100, 'Dummy tasks created successfully.')
+    )
+    @ddt.unpack
+    def test_create_dummy_tasks(self, total, msg):
+        """Test create dummy tasks command."""
+        out = StringIO()
```
Comment (Ayub-Khan) [inline order/management/tests/test_create_dummy_tasks.py]: NIT output instead out out would be better.

File: order/management/tests/test_delete_dummy_orders.py [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+
+class DeleteDummyOrdersTest(TestCase, TestDbSetUp):
+    """Dummy orders deletion test class."""
+
+    def setUp(self):
+        """Setup order for deletion."""
+        self.client = self.create_client(email_post_fix=DUMMY_EMAIL_MARKER)
+
+    def test_delete_dummy_orders(self):
+        """Test delete dummy orders command."""
```
Comment (Ayub-Khan) [inline order/management/tests/test_delete_dummy_orders.py:22]: for all delete tests. add a new line 
self.assertTrue(Order.objects.filter(order=DUMMY_ORDER_MARKER).exists())
as the first line to make sure before we run the command dummy data exists which we are going to delete.

File: order/tests/test_order.py [hunk 25]
```diff
@@ -18,6 +17,7 @@ def setUp(self):
         self.api_client = APIClient()
         self.user, self.username, self.password = self.create_user()
         self.client = self.create_client()
+        _ = self.create_client()
```
Comment (Ayub-Khan) [inline order/tests/test_order.py]: Do we need two clients?

File: order/tests/test_task.py [hunk 26]
```diff
@@ -19,6 +18,7 @@ def setUp(self):
         self.user, self.username, self.password = self.create_user()
         self.client = self.create_client()
         self.order = self.create_order(self.client)
+        _ = self.create_order(self.client)
```
Comment (Ayub-Khan) [inline order/tests/test_task.py]: do we need two orders for tests.

**Review by Ayub-Khan — CHANGES_REQUESTED** [review]

File: employee/management/tests/test_delete_dummy_employees.py (hunk 20 above)
Comment (saadali5997) [reply employee/management/tests/test_delete_dummy_employees.py:18]: It's already using dummy marker in utils, no need to send from here.

File: employee/tests.py [hunk 27]
```diff
@@ -1,4 +1,5 @@
 """Tests for employee application."""
+from io import StringIO
```
Comment (Ayub-Khan) [inline employee/tests.py]: remove this too we are not using this.

File: order/tests/test_order.py [hunk 28]
```diff
@@ -1,6 +1,5 @@
 """Tests for order application."""
-
-from urllib.parse import urlunparse
+from io import StringIO
```
Comment (Ayub-Khan) [inline order/tests/test_order.py]: no need for StringIO i think.

File: order/tests/test_task.py [hunk 29]
```diff
@@ -1,6 +1,5 @@
 """Tests for order application."""
-
-from urllib.parse import urlunparse
+from io import StringIO
```
Comment (Ayub-Khan) [inline order/tests/test_task.py]: no need for stringIo i think.

**Review by Ayub-Khan — APPROVED** [review]

**Pull request merged**